Fix FileSystem related permissions failures.

ianhays opened the pull request:
A number of tests involving ReadOnly files were failing on Unix and Linux when run with elevated user permissions (e.g. sudo). This commit modifies those Unix tests to accommodate either set of permissions and verify the result accordingly. This unfortunately means we can't verify if the behavior is 'right', only that it is executed correctly. Windows is unchanged.

resolves #5039, #5041, #4605

stephentoub commented on src/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs:
```diff
@@ -114,5 +114,31 @@ public void WriteToReadOnlyFile_UnauthException()
+            {
+                File.WriteAllBytes(path, Encoding.UTF8.GetBytes("text"));
+                Assert.Equal(Encoding.UTF8.GetBytes("text"), File.ReadAllBytes(path));
+            }
```
Is there some kind of assert we could add here (and in other similar places) that we're running in an elevated context and that's why it didn't throw? e.g. would it work to check whether geteuid() returns 0?

ianhays replied:
Yes, now that we've split the files into Windows/Unix we can add a geteuid check to the Unix tests. That will solve the Linux case, but I'm not sure what is happening specifically on OSX to be raising errors. I've been working under the assumption that the CI tests are running as the superuser on OSX. If that is incorrect, adding an ```Assert.Equal(0, getuid())``` call would not solve the failures that #4605 found.

ianhays commented:
Updated with feedback and checks for GetEUid on Unix systems. Tested on Ubuntu under sudo and not under sudo and both work.

stephentoub commented:
One question, otherwise LGTM.  Can you also check whether this addresses the issues on OS X?

ianhays commented:
Modified the tests to manually declare the externs. 

>Can you also check whether this addresses the issues on OS X?

It does resolve the issue on OSX, though I was unable to determine why they failed originally.

stephentoub commented on src/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs:
```diff
@@ -104,15 +105,35 @@ public void OpenFile_ThrowsIOException()
+            try
+            {
+                // Operation succeeds when being run by the Unix superuser
+                if ((RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) && geteuid() == 0)
```
Nit: You could make this ```if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && geteuid() == 0)```

stephentoub commented on src/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs:
```diff
@@ -104,15 +105,35 @@ public void OpenFile_ThrowsIOException()
             }
         }
 
-        [ActiveIssue(4605, PlatformID.OSX)]
```
Have we verified this fixes it on OS X?  Since the PR CI isn't currently validating OS X, we'd need to be prepared for this to fail and need a follow-up PR to disable it again.

stephentoub replied:
I just saw you already answered this earlier.  Ignore me :smile:

ianhays replied:
Actually, I'm wrong, I have not verified it on OSX. I didn't realize that the CI wasn't doing OSX verification.

Is there any somewhat simple way to verify without actually having an OSX machine?

stephentoub commented on src/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs:
```diff
@@ -104,15 +105,35 @@ public void OpenFile_ThrowsIOException()
         }
+
+        [DllImport("libc", SetLastError = true)]
+        private static extern int geteuid();
```
Could we move this to some central location, maybe the base class, rather than repeating it multiple times?

stephentoub commented:
A few nits, but otherwise LGTM.

ianhays commented:
@dotnet-bot test this please
